Add hashable instance for StrongPath (#43)

ccntrq opened the pull request:
Hey @Martinsos 

I saw your call for participation in [Issue 300](https://haskellweekly.news/issue/300.html) of the Haskell weekly newsletter.

I went with your suggestion to implement the `hashable` instance by using the `toFilePath` function to resolve Issue #43.
Please let me know if this is what you had in mind or if anything should be added or changed in this PR

Kind regards.

Martinsos commented:
Hey @ccntrq, thanks for doing this, it looks great!

I wasn't aware we will need to separate this into a separate module, but I understand the reasoning, makes sense. Great that you also added a test.

One thing that crossed my mind is -> if we have two relative paths, one relative do Dir A, and another do Dir B, they might have same hash, even though they are in theory different paths. But, I don't see a way around that really, since the fact of what they are relative to is captured in the type system, not at the value level. I assume in practice this should not be a problem, you would probably not mix relative paths that are relative to different things and using their hashes.

I will wait for the CI to pass and then I will merge and release a new version of StrongPath. Thanks again!

ccntrq commented:
Thanks for the encouraging feedback. 

Regarding the possibility of getting the same hash for different paths, I added a note to the instance declaration that will also show up in the generated haddock. 

Added another commits that should make the tests run on windows as well.

Martinsos commented:
Nice, note is a good idea! I don't think there is anything we can do it, it is the nature of the problem, but at least this way we make sure expectations are clear.

Martinsos commented:
@ccntrq unfortunately it is still failing on windows!

The reason is, the assumption that hash of the original FilePath is equal to the hash of the strong path that we get as a result of parsing that FilePath, doesn't always have to be correct. For example, in this case, on Windows, original FilePath has Posix separators, but once you parse it into StrongPath and then try to convert it into FilePath again, it will be printed with Windows separators (since it is a Windows path). This is intended behaviour.

We should design test in a different fashion. Maybe a good way would be to check following properties:
- two different paths give different hash
- path constructed directly from "a/b" has same hash as path that was created by creating two paths that were constructed from "a" and "b" respectively and then concatenated

ccntrq commented:
Good evening @Martinsos 

thanks for the advice. I will have a look into it tomorrow and rework the tests according to your suggestions.

Have a nice evening 🍻

ccntrq commented:
Good morning @Martinsos,

I just finished reworking the tests. I think these are much better tests now than before, since we're not testing an implementation detail anymore. These should always pass, even when we decide to switch out the implementation for something else.

I assume the tests should be passing on windows now as well.

Martinsos commented:
Great @ccntrq , merged, thanks a lot for the effort :)!
